design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,162 @@
+
+![Repository Structure](diagrams/uncommitted-gc-repo-struct.png)
+
+### 2. Listing of lakeFS repository committed objects
+
+Similar to the way GC works today, use repository meta-ranges and ranges to read all committed objects in the repository. 
+
+### 3. Listing of lakeFS repository uncommitted objects
+
+Creating special meta-ranges and ranges using branches' uncommitted data, we can leverage the already existing GC logic to
+quickly read all uncommitted objects in the repository.
+
+### Required changes by lakeFS
+
+The following are necessary changes in lakeFS in order to implement this proposal successfully.
+
+#### Objects Path Conventions
+
+Uncommitted GC must scan the bucket in order to find objects that are not referenced by lakeFS.
+To optimize this process, suggest the following changes:
+
+1. Divide the repository namespace into time-and-size based partitions.
+2. Partition name will be a time based, reverse sorted unique identifier.
+3. LakeFS will create a new partition on a timely basis (for example: hourly) or when it has written < TBD > objects to the partition.
+4. Each partition will be written by a single lakeFS instance in order to track partition size.
+5. The sorted partitions will enable partial scans of the bucket when running the optimized GC.
+
+#### StageObject
+
+The StageObject operation will only be allowed on addresses outside the repository's storage namespace. This way, objects added using this operation are never collected by GC.
+
+#### [Get/Link]PhysicalAddress 
+
+1. GetPhysicalAddress to return a validation token along with the address (or embedded as part of the address).
+2. The token will be valid for a specified amount of time and for a single use.
+3. lakeFS will need to track issued tokens/addresses, and delete them when tokens are expired/used
+4. LinkPhysicalAddress to verify token valid before creating an entry.
+    1. Doing so will allow us to use this time interval to filter objects that might have been uploaded and waiting for
+       the link API and avoid them being deleted by the GC process.
+    2. Objects that were uploaded to a physical address issued by the API and were not linked before the token expired will
+       eventually be deleted by the GC job.
+>**Note:** These changes also solves an existing bug where an issued physical address which was committed can be overwritten by Link
+
+#### S3 Gateway CopyObject
+
+1. Copy object in the same branch will work the same - creating a new staging entry using the existing entry information.
+2. For objects that are not part of the branch, use the underlying adapter copy operation.
+
+#### Move/RenameObject
+Clients working through the S3 Gateway can use the CopyObject + DeleteObject to perform a Rename or Move operation.
+For clients using the OpenAPI this could have been done using StageObject + DeleteObject.
+To continue support of this operation, introduce a new API to rename an object which will be scoped to a single branch.
```

Comment:
It is still possible to use CopyObject + DeleteObject, it will be also possible to use RenameObject.
It will no longer be possible to use StageObject as done in lakeFSFS (renameObject).

In any case you are correct (and I also did not state :) ) that the RenameObject API is atomic

Reply:
I don't understand how can the kv interface support an atomic rename.

Reply:
I'll rephrase - since the parenthesis were confusing.
**RenameObject API is not atomic**

Reply:
If it's not atomic, can we add and use copy instead?

Reply:
@N-o-Z  Still relevant - If we have a way to logically do a rename with copy and delete, why do we need a non-atomic rename in the API?

Reply:
You have a logical way to do it using the S3 Gateway, no way of doing it using the OpenAPI